# Dataset Creation

Combining 3 data sources from the AlphaVantage API into one dataset.

## Import Libraries

In [1]:
import pandas as pd

## Import Data

In [2]:
balance_sheet_data = pd.read_csv('../data/raw/balance_sheet_annual.csv')
income_statement_data = pd.read_csv('../data/raw/income_statement_annual.csv')
eps_data = pd.read_csv('../data/raw/earnings_annual.csv')
overview_data = pd.read_csv('../data/raw/company_overview.csv')

## Quick Checks

### Dataset lengths

In [3]:
print(f"balance sheet data length: {balance_sheet_data.shape}")
print(f"income statement data length: {income_statement_data.shape}")
print(f"company overview data length: {overview_data.shape}")

balance sheet data length: (1539, 39)
income statement data length: (1424, 27)
company overview data length: (101, 52)


The balance sheet data and income statement data length do not match.

### Columns

In [4]:
balance_sheet_data.dtypes

fiscalDateEnding                           object
reportedCurrency                           object
totalAssets                                 int64
totalCurrentAssets                        float64
cashAndCashEquivalentsAtCarryingValue     float64
cashAndShortTermInvestments               float64
inventory                                 float64
currentNetReceivables                     float64
totalNonCurrentAssets                     float64
propertyPlantEquipment                    float64
accumulatedDepreciationAmortizationPPE    float64
intangibleAssets                          float64
intangibleAssetsExcludingGoodwill         float64
goodwill                                  float64
investments                               float64
longTermInvestments                       float64
shortTermInvestments                      float64
otherCurrentAssets                        float64
otherNonCurrentAssets                     float64
totalLiabilities                          float64


In [5]:
income_statement_data.dtypes

fiscalDateEnding                      object
reportedCurrency                      object
grossProfit                            int64
totalRevenue                           int64
costOfRevenue                          int64
costofGoodsAndServicesSold           float64
operatingIncome                        int64
sellingGeneralAndAdministrative      float64
researchAndDevelopment               float64
operatingExpenses                      int64
investmentIncomeNet                  float64
netInterestIncome                    float64
interestIncome                       float64
interestExpense                      float64
nonInterestIncome                    float64
otherNonOperatingIncome              float64
depreciation                         float64
depreciationAndAmortization          float64
incomeBeforeTax                        int64
incomeTaxExpense                       int64
interestAndDebtExpense               float64
netIncomeFromContinuingOperations    float64
comprehens

We can remove the 'reportedCurrency' column and we will merge the two datasets on 'symbol' and 'fiscalDateEnding'. All the other columns are numerical with data type float or integer.

In [6]:
balance_sheet_data.drop('reportedCurrency', axis=1, inplace=True)
income_statement_data.drop('reportedCurrency', axis=1, inplace=True)

### Missing Values

In [7]:
balance_sheet_data.isna().sum()

fiscalDateEnding                             0
totalAssets                                  0
totalCurrentAssets                           6
cashAndCashEquivalentsAtCarryingValue        5
cashAndShortTermInvestments                  6
inventory                                  206
currentNetReceivables                      357
totalNonCurrentAssets                        7
propertyPlantEquipment                      29
accumulatedDepreciationAmortizationPPE     255
intangibleAssets                            69
intangibleAssetsExcludingGoodwill           86
goodwill                                   123
investments                                288
longTermInvestments                        400
shortTermInvestments                       233
otherCurrentAssets                          59
otherNonCurrentAssets                      350
totalLiabilities                             4
totalCurrentLiabilities                      9
currentAccountsPayable                     322
deferredReven

In [8]:
income_statement_data.isna().sum()

fiscalDateEnding                       0
grossProfit                            0
totalRevenue                           0
costOfRevenue                          0
costofGoodsAndServicesSold            42
operatingIncome                        0
sellingGeneralAndAdministrative        2
researchAndDevelopment               510
operatingExpenses                      0
investmentIncomeNet                  614
netInterestIncome                     71
interestIncome                       473
interestExpense                       16
nonInterestIncome                    228
otherNonOperatingIncome              297
depreciation                         475
depreciationAndAmortization            1
incomeBeforeTax                        0
incomeTaxExpense                       0
interestAndDebtExpense               117
netIncomeFromContinuingOperations      6
comprehensiveIncomeNetOfTax           20
ebit                                   0
ebitda                                 1
netIncome       

## Dataset Merge

In [9]:
data = balance_sheet_data.merge(income_statement_data, how='inner', on=['symbol', 'fiscalDateEnding'])
data.shape

(1425, 62)

In [10]:
data.head()

fiscalDateEnding  totalAssets  totalCurrentAssets  \
0       2024-11-02  48228277000        5.484654e+09   
1       2023-10-28  48794478000        4.384022e+09   
2       2022-10-29  50302350000        4.937992e+09   
3       2021-10-30  52322071000        5.378317e+09   
4       2020-10-31  21468603000        2.517688e+09   

   cashAndCashEquivalentsAtCarryingValue  cashAndShortTermInvestments  \
0                           1.991342e+09                 2.363164e+09   
1                           9.580610e+08                 9.580610e+08   
2                           1.470572e+09                 1.470572e+09   
3                           1.977964e+09                 1.977964e+09   
4                           1.055860e+09                 1.055860e+09   

      inventory  currentNetReceivables  totalNonCurrentAssets  \
0  1.447687e+09           1.336331e+09           4.274362e+10   
1  1.642214e+09           1.469734e+09           4.441046e+10   
2  1.399914e+09           1.800462e+09           4.536436e+10   
3  1.200610e+09           1.459056e+09           4.694375e+10   
4  6.082600e+08           7.375360e+08           1.889568e+10   

   propertyPlantEquipment  accumulatedDepreciationAmortizationPPE  ...  \
0            3.415550e+09                            3.772438e+09  ...   
1            3.219157e+09                            3.424775e+09  ...   
2            2.401304e+09                            3.148203e+09  ...   
3            1.979051e+09                            2.956246e+09  ...   
4            1.120561e+09                            2.765095e+09  ...   

   depreciation  depreciationAndAmortization  incomeBeforeTax  \
0   362771000.0                 1.700000e+09       1777340000   
1   334704000.0                 2.000000e+09       3608003000   
2   283338000.0                 2.014200e+09       3098749000   
3   231275000.0                 8.433590e+08       1328714000   
4   233775000.0                 5.771480e+08       1311617000   

   incomeTaxExpense  interestAndDebtExpense  \
0         142067000             322227000.0   
1         293424000             264641000.0   
2         350188000             200408000.0   
3         -61708000             399975000.0   
4          90856000             193305000.0   

   netIncomeFromContinuingOperations  comprehensiveIncomeNetOfTax        ebit  \
0                       1.635273e+09                 1.638319e+09  2032798000   
1                       3.314579e+09                 3.324429e+09  3872644000   
2                       2.748561e+09                 2.736974e+09  3299157000   
3                       1.390422e+09                 1.453318e+09  1513539000   
4                       1.220761e+09                 1.161478e+09  1504922000   

         ebitda   netIncome  
0  3.732798e+09  1635273000  
1  5.872644e+09  3314579000  
2  5.313357e+09  2748561000  
3  2.356898e+09  1390422000  
4  2.082070e+09  1220761000  

[5 rows x 62 columns]

## Add EPS

In [11]:
eps_data.head()

fiscalDateEnding  reportedEPS symbol
0       2024-09-30         1.81    BSX
1       2023-12-31         2.05    BSX
2       2022-12-31         1.75    BSX
3       2021-12-31         1.63    BSX
4       2020-12-31         0.96    BSX

In [12]:
data = data.merge(eps_data, on=['fiscalDateEnding', 'symbol'], how='left')

In [13]:
data.isna().sum()

fiscalDateEnding                           0
totalAssets                                0
totalCurrentAssets                         0
cashAndCashEquivalentsAtCarryingValue      0
cashAndShortTermInvestments                0
                                        ... 
comprehensiveIncomeNetOfTax               20
ebit                                       0
ebitda                                     1
netIncome                                  0
reportedEPS                              142
Length: 63, dtype: int64

We have some missing values in the reportedEPS columns. About 10% of the information is missing and this probably due to the date in fiscalDateEnding being slightly different in both datasets. Let's verify this assumption:

In [14]:
data[data['reportedEPS'].isna()][['fiscalDateEnding', 'symbol']]

fiscalDateEnding symbol
0          2024-11-02    ADI
1          2023-10-28    ADI
2          2022-10-29    ADI
3          2021-10-30    ADI
5          2019-11-02    ADI
...               ...    ...
1166       2015-09-03     MU
1167       2014-08-28     MU
1168       2013-08-29     MU
1169       2012-08-30     MU
1170       2011-09-01     MU

[142 rows x 2 columns]

In [15]:
eps_data[eps_data['symbol'] == 'ADI'].head()

fiscalDateEnding  reportedEPS symbol
1591       2024-10-31         6.38    ADI
1592       2023-10-31        10.08    ADI
1593       2022-10-31         9.59    ADI
1594       2021-10-31         6.43    ADI
1595       2020-10-31         4.91    ADI

Indeed, looking at the symbol ADI, we can see that there is a mismatch between the two datasets about the fiscalDateEnding data was reported and/or recorded. We could try to merge the dataset using the year and month only. However, we want to keep the exact matches so we will create a new column for data merged using only the year and the month. Then we will replace the NaN values from the first operation by the values we get with the second.

In [16]:
data['YM_date'] = data['fiscalDateEnding'].str[:7]
eps_temp = eps_data.copy()
eps_temp['YM_date'] = eps_data['fiscalDateEnding'].str[:7]
eps_temp.drop('fiscalDateEnding', axis=1, inplace=True)

In [17]:
# merge using the newly created YM_date columns
data = data.merge(eps_temp, on=['symbol', 'YM_date'], how='left').drop('YM_date', axis=1)
# replacing the missing values by the new ones
data.loc[data['reportedEPS_x'].isna(), 'reportedEPS_x'] = data.loc[data['reportedEPS_x'].isna(), 'reportedEPS_y']
data.drop('reportedEPS_y', axis=1, inplace=True)

In [18]:
data.isna().sum()

fiscalDateEnding                          0
totalAssets                               0
totalCurrentAssets                        0
cashAndCashEquivalentsAtCarryingValue     0
cashAndShortTermInvestments               0
                                         ..
comprehensiveIncomeNetOfTax              20
ebit                                      0
ebitda                                    1
netIncome                                 0
reportedEPS_x                            50
Length: 63, dtype: int64

We went from 142 missing values to 50. We will now merge the datasets using only the year as a last resort.

In [19]:
data['Y_date'] = data['fiscalDateEnding'].str[:4]
eps_temp = eps_data.copy()
eps_temp['Y_date'] = eps_data['fiscalDateEnding'].str[:4]
eps_temp.drop('fiscalDateEnding', axis=1, inplace=True)

# in case we have several entries of a given symbol and a given year, we will keep the largest EPS which we assume will correspond to full year data
eps_temp = eps_temp.groupby(['Y_date', 'symbol'], as_index=False).max()

In [20]:
# merge using the newly created Y_date columns
data = data.merge(eps_temp, on=['symbol', 'Y_date'], how='inner').drop('Y_date', axis=1)
# replacing the missing values by the new ones
data.loc[data['reportedEPS_x'].isna(), 'reportedEPS_x'] = data.loc[data['reportedEPS_x'].isna(), 'reportedEPS']
data.drop('reportedEPS', axis=1, inplace=True)
data.rename(columns={'reportedEPS_x': 'reportedEPS'}, inplace=True)

In [21]:
data[data['symbol'] == 'LOW']

fiscalDateEnding  totalAssets  totalCurrentAssets  \
75       2024-02-02  41795000000        1.907100e+10   
76       2023-02-03  43708000000        2.144200e+10   
77       2022-01-28  44640000000        2.006000e+10   
78       2021-01-29  46735000000        2.232600e+10   
79       2020-01-31  39471000000        1.531800e+10   
80       2019-02-01  34508000000        1.422800e+10   
81       2018-02-02  35291000000        1.277200e+10   
82       2017-02-03  34408000000        1.200000e+10   
83       2016-01-29  31266000000        1.056100e+10   
84       2015-01-30  31721000000        9.851000e+09   
85       2014-01-31  32732000000        1.029600e+10   
86       2013-02-01  32666000000        9.784000e+09   
87       2012-02-03  33559000000        1.007200e+10   
88       2011-01-28  33699000000        9.967000e+09   
89       2010-01-29  33005000000        9.732000e+09   

    cashAndCashEquivalentsAtCarryingValue  cashAndShortTermInvestments  \
75                           9.210000e+08                 1.228000e+09   
76                           1.348000e+09                 1.732000e+09   
77                           1.133000e+09                 1.404000e+09   
78                           4.690000e+09                 5.196000e+09   
79                           7.160000e+08                 8.760000e+08   
80                           5.110000e+08                 7.290000e+08   
81                           5.880000e+08                 6.900000e+08   
82                           5.580000e+08                 6.580000e+08   
83                           4.050000e+08                 7.120000e+08   
84                           4.660000e+08                 5.910000e+08   
85                           3.910000e+08                 5.760000e+08   
86                           5.410000e+08                 6.660000e+08   
87                           1.014000e+09                 1.300000e+09   
88                           6.520000e+08                 1.123000e+09   
89                           6.320000e+08                 1.057000e+09   

       inventory  currentNetReceivables  totalNonCurrentAssets  \
75  1.689400e+10                    NaN           5.748000e+09   
76  1.853200e+10                    NaN           2.236600e+10   
77  1.760500e+10                    NaN           2.676800e+10   
78  1.619300e+10                    NaN           2.545900e+10   
79  1.317900e+10                    NaN           2.508000e+10   
80  1.256100e+10                    NaN           2.024200e+10   
81  1.139300e+10                    NaN           2.290000e+10   
82  1.045800e+10                    NaN           2.276200e+10   
83  9.458000e+09                    NaN           2.090700e+10   
84  8.911000e+09                    NaN           2.205200e+10   
85  9.127000e+09                    NaN           2.270400e+10   
86  8.600000e+09                    NaN           2.314500e+10   
87  8.355000e+09                    NaN           2.374900e+10   
88  8.321000e+09                    NaN           2.399200e+10   
89  8.249000e+09                    NaN           2.327300e+10   

    propertyPlantEquipment  accumulatedDepreciationAmortizationPPE  ...  \
75            1.765300e+10                                     NaN  ...   
76            1.756700e+10                            1.734400e+10  ...   
77            1.907100e+10                            1.788800e+10  ...   
78            1.915500e+10                            1.754700e+10  ...   
79            1.876900e+10                            1.731000e+10  ...   
80            1.843200e+10                            1.743100e+10  ...   
81            1.972100e+10                            1.721900e+10  ...   
82            1.994900e+10                            1.696900e+10  ...   
83            1.957700e+10                            1.633600e+10  ...   
84            2.003400e+10                            1.540900e+10  ...   
85            2.083400e+10             

In [22]:
data.isna().sum()

fiscalDateEnding                          0
totalAssets                               0
totalCurrentAssets                        0
cashAndCashEquivalentsAtCarryingValue     0
cashAndShortTermInvestments               0
                                         ..
comprehensiveIncomeNetOfTax              20
ebit                                      0
ebitda                                    1
netIncome                                 0
reportedEPS                               0
Length: 63, dtype: int64

Finally, we have no more missing values for the reportedEPS column. To do so we first try to match the dates exactly which enabled us to recover about 90% of the EPS data. We then recover 7% using only the year and the month as a match. Finally, for the remaining 3% we matched the data using the year only.

## Add Sector & Industry Information

In [23]:
overview_data = overview_data[['symbol', 'Sector', 'Industry']]
overview_data.head()

symbol            Sector                                         Industry
0   TMUS        TECHNOLOGY                    RADIOTELEPHONE COMMUNICATIONS
1     PM     MANUFACTURING                                       CIGARETTES
2     GS           FINANCE  SECURITY BROKERS, DEALERS & FLOTATION COMPANIES
3   AAPL        TECHNOLOGY                             ELECTRONIC COMPUTERS
4      V  TRADE & SERVICES                  SERVICES-BUSINESS SERVICES, NEC

In [24]:
data = data.merge(overview_data, on='symbol', how='left')
data.head()

fiscalDateEnding  totalAssets  totalCurrentAssets  \
0       2024-11-02  48228277000        5.484654e+09   
1       2023-10-28  48794478000        4.384022e+09   
2       2022-10-29  50302350000        4.937992e+09   
3       2021-10-30  52322071000        5.378317e+09   
4       2020-10-31  21468603000        2.517688e+09   

   cashAndCashEquivalentsAtCarryingValue  cashAndShortTermInvestments  \
0                           1.991342e+09                 2.363164e+09   
1                           9.580610e+08                 9.580610e+08   
2                           1.470572e+09                 1.470572e+09   
3                           1.977964e+09                 1.977964e+09   
4                           1.055860e+09                 1.055860e+09   

      inventory  currentNetReceivables  totalNonCurrentAssets  \
0  1.447687e+09           1.336331e+09           4.274362e+10   
1  1.642214e+09           1.469734e+09           4.441046e+10   
2  1.399914e+09           1.800462e+09           4.536436e+10   
3  1.200610e+09           1.459056e+09           4.694375e+10   
4  6.082600e+08           7.375360e+08           1.889568e+10   

   propertyPlantEquipment  accumulatedDepreciationAmortizationPPE  ...  \
0            3.415550e+09                            3.772438e+09  ...   
1            3.219157e+09                            3.424775e+09  ...   
2            2.401304e+09                            3.148203e+09  ...   
3            1.979051e+09                            2.956246e+09  ...   
4            1.120561e+09                            2.765095e+09  ...   

   incomeTaxExpense  interestAndDebtExpense  \
0         142067000             322227000.0   
1         293424000             264641000.0   
2         350188000             200408000.0   
3         -61708000             399975000.0   
4          90856000             193305000.0   

   netIncomeFromContinuingOperations  comprehensiveIncomeNetOfTax        ebit  \
0                       1.635273e+09                 1.638319e+09  2032798000   
1                       3.314579e+09                 3.324429e+09  3872644000   
2                       2.748561e+09                 2.736974e+09  3299157000   
3                       1.390422e+09                 1.453318e+09  1513539000   
4                       1.220761e+09                 1.161478e+09  1504922000   

         ebitda   netIncome  reportedEPS         Sector  \
0  3.732798e+09  1635273000         6.38  MANUFACTURING   
1  5.872644e+09  3314579000        10.08  MANUFACTURING   
2  5.313357e+09  2748561000         9.59  MANUFACTURING   
3  2.356898e+09  1390422000         6.43  MANUFACTURING   
4  2.082070e+09  1220761000         4.91  MANUFACTURING   

                           Industry  
0  SEMICONDUCTORS & RELATED DEVICES  
1  SEMICONDUCTORS & RELATED DEVICES  
2  SEMICONDUCTORS & RELATED DEVICES  
3  SEMICONDUCTORS & RELATED DEVICES  
4  SEMICONDUCTORS & RELATED DEVICES  

[5 rows x 65 columns]

## Saving the dataset

In [25]:
data.to_csv('../data/preprocessed/preprocessed_data.csv', index=False)

## What's next?

Next step consists in building the target variable. Although we brought in the dataset the EPS (earning per share) information for the correspondiong date, what we actually need as a target is the EPS value x years (5, 10) from the sample date.